[C-3193] Harmony Text Input

DejayJD opened the pull request:
### Description

* Ported `InputV2` to Harmony now named `TextInput` (along with it's companions `HelperText` and `useFocusState`)
* Left `InputV2` intact for now, we can refactor later
* Add `error` to `TextColor` list
* Refactored the existing sizes and variant to align closer to design
* Add start/end icon support

### How Has This Been Tested?

Lots of manual poking around Storybook

DejayJD commented on packages/harmony/src/components/input/TextInput/HelperText.tsx:
```diff
@@ -0,0 +1,21 @@
+}
+
+export const HelperText = (props: HelperTextProps) => {
+  const { children, hasError } = props
```
renamed `error` to `hasError` here since its just a boolean

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,206 @@
+import styles from './TextInput.module.css'
+import { useFocusState } from './useFocusState'
+
+export enum TextInputSize {
+  SMALL = 'small',
+  DEFAULT = 'default'
+}
```
Previously there was a "large" size but design wants to stick to just 2 moving forward (the large was only like a few pixels bigger anyways)

DejayJD commented on packages/harmony/src/components/typography/Text/types.ts:
```diff
@@ -22,7 +22,7 @@ export type TextProps<TextComponentType extends ElementType = 'p'> =
 export type TextStrength = 'weak' | 'default' | 'strong'
 
-export type TextColor = 'heading' | 'default' | 'subdued' | 'disabled'
+export type TextColor = 'heading' | 'default' | 'subdued' | 'disabled' | 'error'
```
@Kyle-Shanks Added the `error` text color 👀

Kyle-Shanks replied:
nice! `danger` or something more cover more cases like destructive button text which isn't an error, but this works.

DejayJD replied:
I just followed whatever Sammie did on the design side, just asked about `danger` instead
https://audius-internal.slack.com/archives/C05MUGKK4AG/p1697499291702829?thread_ts=1697129541.855729&cid=C05MUGKK4AG

DejayJD replied:
This was updated to `danger` on both the design side and here 👍

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+   * Hides the label. If the label is hidden the placeholder will show instead.
+   * @default false
+   */
+  hideLabel?: boolean
```
Previously this was being used as the "variant", but the only behavior between the two was label or no label so I decided to make a toggle option instead

audius-infra commented:
Preview this change https://demo.audius.co/jd/harmony-text-input

Kyle-Shanks commented on packages/harmony/src/components/layout/layout.module.css:
```diff
@@ -0,0 +1,34 @@
+.row {
+    display: flex;
+    flex-direction: row;
+    align-items: center;
+  }
+  
+  .col {
+    display: flex;
+    flex-direction: column;
+  }
+  
+  .gap1 {
+    gap: var(--unit-1);
+  }
+  .gap2 {
+    gap: var(--unit-2);
+  }
+  .gap3 {
+    gap: var(--unit-3);
+  }
+  .gap4 {
+    gap: var(--unit-4);
+  }
+  .gap5 {
+    gap: var(--unit-5);
+  }
+  .gap6 {
+    gap: var(--unit-6);
+  }
+  
+  .gapAuto {
+    justify-content: space-between;
+  }
+  
```
can we try using the Flex component to cover these cases instead of classes?

DejayJD replied:
Oh yes ty. Great callout. I started these changes before you merged yours haha so I'll just remove these and adjust accordingly

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+.root {
+  display: flex;
+  flex-direction: column;
+  width: 100%;
+}
```
could use `<Flex>` here and elsewhere

amendelsohn replied:
Adopt `<Flex />` here and elsewhere

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+.inputRoot {
+  display: flex;
+  align-items: center;
+  gap: var(--harmony-unit-3);
+  padding: var(--harmony-unit-4);
```
gap and padding should use the `--harmony-spacing` values

amendelsohn replied:
```suggestion
  gap: var(--harmony-spacing-m);
  padding: var(--harmony-spacing-l);
```

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+  gap: var(--harmony-unit-3);
+  padding: var(--harmony-unit-4);
+  border: 1px solid var(--harmony-border-default);
+  border-radius: var(--harmony-unit-1);
```
`--harmony-border-radius-s`

amendelsohn replied:
```suggestion
  border-radius: var(--harmony-border-radius-s);
```

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+}
+
+.large {
+  --gap: var(--harmony-unit-1);
```
```suggestion
  --gap: var(--harmony-spacing-xs);
```

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+}
+
+.root:hover {
+  border-color: var(--harmony-neutral-light-6);
```
```suggestion
  border-color: var(--harmony-n-200);
```

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+}
+
+.disabled {
+  background-color: var(--harmony-neutral-light-9);
+  border-color: var(--harmony-neutral-light-8);
```
```suggestion
  background-color: var(--harmony-n-50);
  border-color: var(--harmony-n-100);
```

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+  outline: 0;
+  border: 0;
+  background: none;
+  font-size: var(--harmony-font-size);
```
I don't think this exists. We have a variety of sizes like `--harmony-font-xs`, etc.
(leaving one comment but there are other references to this)

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+
+.textInput::placeholder,
+.label {
+  color: var(--harmony-neutral-light-4);
```
```suggestion
  color: var(--harmony-n-400);
```

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+.label {
+  color: var(--harmony-text-subdued);
+  position: absolute;
+  z-index: 2;
```
We may want to be careful about z-indexes. Is there risk of this being hidden under or showing through other elements in the app if they have z-index specified?

DejayJD replied:
Great callout. I actually ran into this when we made our text input at Bestow. 
I added `isolation: isolate` to the root so that the z-index doesn't spill over anywhere
https://developer.mozilla.org/en-US/docs/Web/CSS/isolation

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,147 @@
+  position: relative;
+  padding-top: var(--harmony-unit-2);
+  align-self: flex-start;
+  color: var(--harmony-neutral-light-4);
```
```suggestion
  color: var(--harmony-n-400);
```

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+  /**
+   * Show a "x/Max" text displaying number of characters allowed
+   */
+  showMaxLength?: boolean
```
I remember design had a request to be able to show this conditionally when you get close to the end. Don't need to add scope here, but flagging anyway

DejayJD replied:
I met up with Sammie and re-did this, now it has a threshold for showing and a threshold for changing to the warning color

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+    const label = required ? `${labelProp} *` : labelProp
+
+    /**
+     * Since Firefox doesn't support the :has() pseudo selector,
```
OOF

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+  DEFAULT = 'default'
+}
+
+export type TextInputProps = Omit<
```
move this to input/types.ts

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+
+import cn from 'classnames'
+
+import layoutStyles from 'components/layout/layout.module.css'
```
don't think layoutStyles exists in harmony anymore

DejayJD replied:
Yeah I started this PR before the layout stuff. Yoinked this in favor of the layout stuff of Flex components and layout tokens

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+            disabled={disabled}
+            placeholder={shouldShowPlaceholder ? placeholder : undefined}
+            aria-label={
+              label ?? endAdornmentText ?? startAdornmentText ?? placeholder
```
are the adornments more appropriate than the placeholder? maybe `label ?? startAdornment + placeholder + endAdornment`?

DejayJD replied:
Changed this behavior, see my other comment https://github.com/AudiusProject/audius-protocol/pull/6329#discussion_r1364280814

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+            inputRender
+          )}
+
+          {showMaxLength && (
```
per our [style guide](https://www.notion.so/audiusproject/Javascript-Typescript-Code-Style-4f34640139fc4a809c3765131bc597ff?pvs=4) which I totally forgot to share with you, use `condition ? <></> : null`

dylanjeffers commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,213 @@
+}
+
+export const TextInput = forwardRef<HTMLInputElement, TextInputProps>(
+  (props: TextInputProps, ref) => {
```
techincally dont need to put the type on props here, but maybe you do to make harmony work? also can you double check in the react dev-tools if text-input gets named collectly with forwardRef? what weve don in the past is do `const A = forwardRef(function A() {...})`

DejayJD replied:
Checked this and the name is still showing fine in devtools (at least in storybook)

dylanjeffers replied:
Awesome, thanks. Just realized the place to check is in web after the build. I think it might get lost at that point?

amendelsohn reviewed (approved):
Excellent work! Approved, pending the change to layout and styling.

dylanjeffers reviewed (approved):
looks excellent! agreed with kj's comments!

DejayJD commented:
Bunch of new changes up for re-reviewing!
Some highlights:
- Made label wrap the entire thing so that focus works correctly (paired with `useId()` to make sure it always works/is accessible)
- Reworked centering logic so that adornments always stay in line with the input content and the maxlength text stays in line with the label
- Adornments now only show on focus/hasValue
- Added z-index isolation so that it doesnt leak out
- Reworked maxLength text logic. New behavior is that there are two percentage threshold numbers, one for when it decides to appear (70% by default) and one for when it turns orange (90% by default). Sammie and I discussed this one a bunch and this is what she liked the most. It also only shows when focussed in the input
  - Also, we left the "orange border" warning state in, but the original use case for the orange border was just for text
- Fixed font sizes not matching designs
- Added disabled styles
- Separated the `types.ts` file
- Refactored out containers in favor of `Flex`
- Fixed all the harmony tokens to use appropriate things
- Generally added comments and hopefully cleaned up the class names to make more sense (imo)

DejayJD commented on packages/harmony/src/components/input/TextInput/types.ts:
```diff
@@ -0,0 +1,86 @@
+   * @default 0.9
+   */
+  maxLengthWarningThreshold?: number
+} & ( // Make either label or aria-label required prop
+    | {
+        /**
+         * Label Text (Required if aria-label is not provided)
+         */
+        label: string
+        ['aria-label']?: string
+      }
+    | {
+        /**
+         * Label Text (Required if aria-label is not provided)
+         */
+        ['aria-label']: string
+        label?: string
+      }
+  )
```
Would love some feedback on this, put this in here as kind of a "proposal". 
This effectively makes providing either `label={}` or `aria-label={}` required. A bit heavy handed but makes sure we always solve for accessibility. 
The only part that I really don't like is that the TS error isn't very helpful which might make this not worth 😅 
I'm open to whatever others think
![image](https://github.com/AudiusProject/audius-protocol/assets/6711655/2c8a78f8-b10e-44c5-a14d-b7579d0f264c)

dylanjeffers replied:
I like your thinking, my hope is that we assign this to the size of the input, like if it's a small input, aria-label is required, and for large label is required and aria-label optional or maybe not allowed?

amendelsohn replied:
Ooh this is a tough one. I'm tempted to say we should just require label and put a comment on the type saying it's required for accessibility

dylanjeffers replied:
ahh so if you provide label, then it becomes the aria-label for the smaller input? i like that a lot

DejayJD replied:
Yeah I agree, I'm gonna do `label` required regardless of whether it's shown visually or not. 
I thought it was being slick with that union setup until I saw that error message 😂

rickyrombo replied:
This might be helpful here! https://gist.github.com/palashmon/db68706d4f26d2dbf187e76409905399

audius-infra commented:
Preview this change https://demo.audius.co/jd/harmony-text-input

dylanjeffers commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+}
+
+/** Shrink the size of the label **/
+.focused .innerRootContainer .label,
```
Is there a way to not depend on specificity in css and make it explicit with a className on the element? Is it cause focused is not being captured in js?

rickyrombo replied:
Fwiw this probably was inherited from my approach in InputV2. I actually prefer this approach vs manually tracking state and classes for everything in JS (css awesome!) but I understand if using JS is an easier approach for folks

DejayJD replied:
Yeah I cant remember why I ran into a specificity issue here but I replaced it with `.label.focused` instead

dylanjeffers commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,178 @@
+  )
+})
+
+export { TextInput }
```
We typically do export const TextInput at the time of definition, is this better?

DejayJD replied:
No difference, I usually prefer exporting at time of definition too. This was actually just leftover when I had `TextInput.displayName = "TextInput"`. I'll move it back

amendelsohn commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,178 @@
+        maxLength={maxLength}
+        disabled={disabled}
+        placeholder={shouldShowPlaceholder ? placeholder : undefined}
+        aria-label={labelText ?? props['aria-label']}
```
if someone provides `label` and `aria-label`, shouldn't we use the `aria-label` first?

DejayJD replied:
Made label required instead, so no longer need to mess with this

dylanjeffers reviewed (approved):
Excellent work!!! Avail to chat about any of my comments, but high level what you've done here is really wonderful

amendelsohn reviewed (approved):
Excellent work! Very clean and well documented.
Lots of tough problems and edge cases here, all very well handled!

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+}
+
+/** Adds the "*" to required fields, but have it disappear when elevated **/
+.required:not(.focused) .innerRootContainer .placeholder:not(.hasValue)::after {
```
Oh yeah I actually didn't change this bit from InputV2, but now that I'm looking at this more closely I'm not loving having the * on both label/placeholder. I'll ask Sammie if she has a preference, but I'm thinking we only show the asterisk on placeholder when the label is hidden?

dylanjeffers replied:
Yeah ask her and I agree just placeholder if no label

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+.disabled {
+  cursor: not-allowed;
+  & * {
+    cursor: not-allowed;
```
This is usually my strategy since the whole thing is normally clickable and theres some z-index stuff going on

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+}
+
+/* disabled styles */
+.disabled {
```
My understanding was that `cursor: not-allowed` is the general best practice for disabled things for a little added. The added context is nice and doesn't prevent you from doing anything like selecting the text or anything
![2023-10-18 17 59 59](https://github.com/AudiusProject/audius-protocol/assets/6711655/077ef2d2-b84c-4a9d-8336-15cc467ff823)

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+/* Actual input element */
+.inputElement {
+  padding: 0;
+  overflow: visible;
```
Oh yeah just checked and actually it is, I thought that `auto` was the default.

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+  background: none;
+  font-weight: var(--harmony-font-medium);
+  color: var(--harmony-text-default);
+  text-overflow: ellipsis;
```
Yeah I actually forgot that I threw this in there 😅 , meant to ask for feedback on it. I also need to see if design has a preference. Since by default `input` doesn't do any form of multiline wrapping the default behavior just cuts off.
Personally I would advocate for ellipsis cause it gives a better visual indicator that there is more text than the "cut off".
Here's with ellipsis
![2023-10-18 18 19 58](https://github.com/AudiusProject/audius-protocol/assets/6711655/f9961d13-4632-40e7-ad26-b392a29f34f4)
Without ellipsis
![2023-10-18 18 21 24](https://github.com/AudiusProject/audius-protocol/assets/6711655/0402e34d-8c16-41f2-bf1b-ce4f4748fd4f)

DejayJD replied:
Side note, this made me check the label too and I think we could get away with growing the height with the container using a min-height instead of height, so now I did this
![image](https://github.com/AudiusProject/audius-protocol/assets/6711655/99dd63e3-83c2-4ee9-9642-78e42cd78820)

amendelsohn replied:
ohh the ellipsis is only visible when unfocused. That's much more reasonable. Nice

Idk how I feel about the multiline label. I'd defer to design

DejayJD replied:
Design said ellipsis overflow on both, so we're good there 👍

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.stories.tsx:
```diff
@@ -0,0 +1,95 @@
+    label: 'Input Label',
+    helperText: 'This is assistive text.',
+    disabled: true,
+    value: 'You couldnt change me if you tried'
```
![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExa3I3c3ZsdzA4NWdud2Zxcmt5amg0NHUycWFoZmsxbHZnaDRrcmN0MiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/O4EChIxazzrHi/giphy-downsized.gif)

rickyrombo commented on packages/harmony/src/components/input/TextInput/HelperText.module.css:
```diff
@@ -0,0 +1,3 @@
+.root {
+  margin-top: var(--harmony-spacing-xs);
```
This feels weird? Doesn't the root input have a flex gap to handle this?

DejayJD replied:
There wasn't a gap before but I definitely prefer that so I moved it over and yoinked this whole component

rickyrombo commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,178 @@
+        </div>
+
+        {EndIcon ? <EndIcon /> : null}
+        {children}
```
Do we want children to render inside the label? That feels somewhat odd to me as we could explicitly specify a label prop and have the actual label be different

rickyrombo replied:
It might have been for the helper text? Or icons? I don't remember 🤦‍♂️

rickyrombo commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,178 @@
+  return (
+    <Flex className={cn(styles.root, className)} direction='column'>
+      <label
+        htmlFor={id}
```
We don't need htmlFor if the label wraps the input, right?

DejayJD replied:
While technically true, an explicit `for` is a best practice for accessibility since some screen readers can't handle implicit labels very well. 
Given that it's easy for us to do an explicit label here I don't see why we wouldn't 🤷 
![image](https://github.com/AudiusProject/audius-protocol/assets/6711655/73fd86e3-bca8-4218-b40b-1efc3eddf8fc)
https://www.w3.org/WAI/tutorials/forms/labels/#associating-labels-implicitly

DejayJD replied:
oof, fun. I didn't realize wrapping was problematic too 💀 
Not wrapping is trickier to implement though since we need the labels to wrap the "clickable" parts for focus management. Otherwise we would have to set up our own click handler to redirect focus.
Since a wrapped+explicit label "still works" according to them, I'm kind of inclined to just leave it as-is. Thoughts?

rickyrombo replied:
SGTM

rickyrombo commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,178 @@
+  )
+
+  // For focus behavior and accessiblity, <label> needs to have a htmlFor={} provided to an id matching the input
+  const backupId = useId()
```
Aren't we wrapping the label around the input? I don't think this is necessary

DejayJD replied:
Same answer as above. It's an accessibility best practice

rickyrombo commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,178 @@
+                  tag='span'
+                  color={showMaxlengthWarningColor ? 'warning' : 'default'}
+                >
+                  {characterCount}/{maxLength}
```
Do we want this as part of the label? (I'm thinking we probably do, but making sure!)

DejayJD replied:
Visually yes, we want it to stay aligned with the label. 
It won't actually be read out  to a screen reader as part of the label since we're providing label via `aria-label`
(from Figma)
![image](https://github.com/AudiusProject/audius-protocol/assets/6711655/9fe089ef-5d3a-4a0c-bab4-68a7f962f72f)

rickyrombo replied:
Oh sorry, I missed that were always setting `aria-label`

Though I think if we went the non-wrapped label route maybe we can avoid `aria-label` in the non `hideLabel` case?

rickyrombo commented on packages/harmony/src/components/input/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,178 @@
+        {EndIcon ? <EndIcon /> : null}
+        {children}
+      </label>
+      {helperText ? (
```
This feels like _maybe_ it should be left to the caller? I tend to lean towards composability vs coupling this sort of thing. In this case, maybe a render fn with `hasError` and the caller can define the text component?

The way it is here is convenient as a dev as they'll likely be coupled a lot, so I can see it both ways, but I often find limiting this sort of thing to a string bites us sometimes when we want to add eg icons or whatever else in a future special case.

This might make sense to be the children prop, too, and have that children prop be a render fn with the error state.

DejayJD replied:
Imo, I don't see much benefit of making it more composable here. If this piece was composable the consumer would have to handle color, size, and gap spacing which all come from the harmony figmas. Plus, I think the majority of cases will be wanting the "standard behavior" anyways; and if it's non-standard we most likely have some room to push back on designs or we can revisit this implementation then.
The other reason is that there's technically already an "escape hatch". If no `helperText` is provided, it's not rendered so if a special case does arise, the composer still has a way to manage it. Honestly this already feels like the best of both worlds; we get a quick & easy access for the majority of cases and an DIY option for special cases.

Also, fwiw, I personally lean somewhere in the middle between composability vs abstracted; maybe a bit more biased towards abstracted. My opinion with design system components is that they should "quickly solve for standard use cases, but not restrict special cases (ideally)". 
With the design system at Bestow, I actually started out with components more heavily favoring composability and found that most of the time it was unnecessary extra effort; but at worst created refactoring headaches if things changed on the design system side. 
Also I generally find it easier to start more abstracted and peel back layers as situations arise. Also in a future state when the majority of designs are "harmonized" there's less reasons to deviate from the standard stuff anyways

Happy to discuss further though

rickyrombo replied:
SGTM! Was a super nitpick. A happy medium here might be typing the prop as a `ReactNode` and leaving everything else, but honestly I'm fine with how you have it and we can always revisit!

DejayJD replied:
Oh yeah I actually like doing that regardless tbh. I updated the type

rickyrombo commented on packages/harmony/src/components/input/TextInput/HelperText.tsx:
```diff
@@ -0,0 +1,27 @@
+  return (
+    <div className={styles.root}>
+      <Text
+        variant='body'
```
Are these props mostly defaults? Feels like maybe we don't need a new component for this?

DejayJD replied:
Yeah looking at it now, I'm down to delete this as an explicit component. You're right in that there's not much happening in this component and some of it is happening in TextInput anyways (size, for one)

rickyrombo reviewed (commented):
Awesome to see InputV2 getting ported!! Great job!

Sorry for the late night review! I'm in Amsterdam right now and reviewed on my phone. I did the `InputV2` component so figured I'd have some context.

Had a couple comments, mostly around the `HelperText` component regarding coupling/composability, lmk what you think

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+}
+
+/* The row containing the input + adornments */
+.inputAdornmentRow {
+  width: 100%;
+  justify-content: space-between;
+}
+
+/** 
+ * Container inbetween the icons, contains the label + input + maxlength text + adornments
+ **/
+.innerRootContainer {
+  position: relative;
+  width: 100%;
+  height: 100%;
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  gap: var(--harmony-spacing-xs);
```
Yeah actually it does! I replaced this with a flex 💯

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+  align-items: center;
+  /* Dont need Y padding since the flex centering takes care of it */
+  padding: 0 var(--harmony-spacing-l);
+  /* The gap spaces the content around the icons */
```
Yea this was accidental

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+  align-items: center;
+  /* Dont need Y padding since the flex centering takes care of it */
+  padding: 0 var(--harmony-spacing-l);
+  /* The gap spaces the content around the icons */
```
Oops, accidental

DejayJD commented on packages/harmony/src/components/input/TextInput/TextInput.module.css:
```diff
@@ -0,0 +1,149 @@
+}
+
+/* The inputRoot is the root container of all things that aren't assistive text */
+.inputRoot {
```
After removing some additional containers due to `Flex`, I ended up with 
* `.root`
* `.contentContainer` (the thing in question here)
* `.input`
* `.label`

Honestly shoutout to @Kyle-Shanks's `Flex` doing great here; it got rid of all the `innerOuterRootContainerWrapper` junk 💯
I think it removed like at least 3 or 4 container classes in this case

audius-infra commented:
Preview this change https://demo.audius.co/jd/harmony-text-input

audius-infra commented:
Preview this change https://demo.audius.co/jd/harmony-text-input

audius-infra commented:
Preview this change https://demo.audius.co/jd/harmony-text-input

jpangelle reviewed (commented):
nice work!!
